dask-gateway/dask_gateway/client.py

```diff
@@ -372,7 +372,7 @@ def __exit__(self, *args):
         self.close()
 
     def __del__(self):
-        if (
+        if hasattr(self, "_asynchronous") and (
```

Comment:
In case the property changes implementation, should maybe check the referenced name instead of an implementation detail.

```suggestion
        if hasattr(self, "asynchronous") and (
```

Reply:
Let me check that. I assumed hasattr always returned true for properties, but maybe not if they raise an AttributeError?

Reply:
OK switched to use `asynchronous` and there's no warning, but I have no idea why it's working. It seems like `hasattr` does always evaluate true in this case:

```python
In [1]: class Foo:
   ...:     @property
   ...:     def a(self):
   ...:         return self._a
   ...:

In [2]: hasattr(Foo, "a")
Out[2]: True
```

so I would expect the warning to be back when using the property name. But I guess it doesn't really matter, as long as it works.

Reply:
You're checking on the class above (which will always work), when what you meant to do was check on an instance

```python
>>> hasattr(Foo, "a")
True
>>> hasattr(Foo(), "a")
False
```

`hasattr` is basically:

```python
def hasattr(x, name):
    try:
        getattr(x, name)
        return True
    except AttributeError
        return False
```